include/deal.II/grid/grid_refinement.h

```diff
@@ -48,6 +48,41 @@ template <class T> class Vector;
 namespace GridRefinement
 {
   /**
+   * This function adjust cell number fraction of refinement and coarsening
+   * according to current cell number and maximal cell number.
+   *
+   * @arg @p current_n_cells is current cell number.
+   *
+   * @arg @p max_n_cells is specified a maximal number of cells. If current
+   * cell number is already exceeded maximal cell number, @p max_n_cells will
+   * be set to zero and @p bottom_fraction will be adjusted to reduce cell
+   * number to @ max_n_cells. If cell number is going to be exceeded only
+   * upon refinement, then refinement and coarsening fractions are going
+   * to be adjusted with a same ratio in an attempt to reach the maximum
+   * number of cells. Be aware though that through proliferation of refinement
+   * due to Triangulation::MeshSmoothing, this number is only an indicator.
+   * The default value of this argument is to impose no limit on the number
+   * of cells.
+   *
+   * @arg @p top_fraction is required refinement fraction. It's value may be
+   * changed after this function call.
+   *
+   * @arg @p bottom_fraction is required coarsening fraction. It's value may be
+   * changed after this function call.
+   *
+   * @note Usually you do not need to call this function explicitly. Pass
+   * @p max_n_cells to function refine_and_coarsen_fixed_number or function
+   * refine_and_coarsen_fixed_fraction and they will call this function if
+   * necessary.
+   */
+  template <int dim>
+  void
+  adjust_refine_and_coarsen_number_fraction (const unsigned int  current_n_cells,
+                                             const unsigned int  max_n_cells,
+                                             double             &top_fraction,
+                                             double             &bottom_fraction);
```

Comment:
Using reference leads to a much more clean code than using std::pair<double,double>.

If return modified fractions with std::pair, then you will need to prepare a temporary std::pair to receive the return value, and then extract the values from std::pair to double variables in order to pass them to flowing functions.

After all, If you still insist on using std::pair<double,double>  with these consideration, I will update the code as your wish.